### PR TITLE
fixed concurrency issues, interleaving issue and other bugs

### PR DESCRIPTION
**Pull request by shreyakhajanchi**
Bugs fixed:

1. Added locking on conv.SchemaIssues to remove concurrent map read/write issues
2. Added shard id column to foreign key in case a rule is added to add shard id column t PK
3. All unsupported data types - should be made nullable by default in HB - Eg Geometry. show a warning/note for the same
4. Skip foreign key creation should be disabled when choosing 'Low downtime migration'
5. Usability - filter should mention this applies only to tables
6. Convert to interleaved - does not account for PK change
7. Convert to interleaved - After PK change, convert to interleaved does not check for order of primary keys
8. Change in order of PK in parent table - does not change order in child table
9. Change order of interleaving in child. Works for first 2 keys, but not for added columns in PK. Causes 2 child tables to appear
10. Deleting interleaved table - does not throw up warning / error. Child table interleaving broken
11. Can not restore interleaved parent - error

### CODE REVIEW

File: internal/convert.go [hunk 1]
```diff
@@ -28,25 +29,26 @@ import (
...
-	UniquePKey     map[string][]string // Maps Spanner table name to unique column name being used as primary key (if needed).
-	Audit          Audit               `json:"-"` // Stores the audit information for the database conversion
-	Rules          []Rule              // Stores applied rules during schema conversion
-	IsSharded      bool                // Flag denoting if the migration is sharded or not
+	mode             mode                     // Schema mode or data mode.
+	SpSchema         ddl.Schema               // Maps Spanner table name to Spanner schema.
+	SyntheticPKeys   map[string]SyntheticPKey // Maps Spanner table name to synthetic primary key (if needed).
+	SrcSchema        map[string]schema.Table  // Maps source-DB table name to schema information.
+	SchemaIssues     map[string]TableIssues   // Maps source-DB table/col to list of schema conversion issues.
+	SchemaIssuesLock sync.RWMutex
```
[the remark this answers is not in the document]
Comment (shreyakhajanchi) [reply internal/convert.go]: added a TODO

File: sources/common/toddl.go [hunk 2]
```diff
@@ -106,10 +106,19 @@ func SchemaToSpannerDDLHelper(conv *internal.Conv, toddl ToDdl, srcTable schema.
 		if len(issues) > 0 {
 			columnLevelIssues[srcColId] = issues
 		}
+		// Set the not null constraint to false for unsupported source datatypes
+		isNotNull := srcCol.NotNull
+		if findSchemaIssue(issues, internal.NoGoodType) != -1 {
+			isNotNull = false
+		}
+		// Set the not null constraint to false for array datatype
```
[the remark this answers is not in the document]
Comment (shreyakhajanchi) [reply sources/common/toddl.go]: added the comment

File: internal/reports/report_helpers.go [hunk 3]
```diff
@@ -81,15 +81,18 @@ func buildTableReport(conv *internal.Conv, tableId string, badWrites map[string]
 		issues, cols, warnings := AnalyzeCols(conv, tableId)
 		tr.Cols = cols
 		tr.Warnings = warnings
-		tr.Errors = int64(len(conv.SchemaIssues[tableId].TableLevelIssues))
+		conv.SchemaIssuesLock.RLock()
+		schemaIssues := conv.SchemaIssues[tableId].TableLevelIssues
```
[the remark this answers is not in the document]
Comment (shreyakhajanchi) [reply internal/reports/report_helpers.go:85]: We have added read and write locks at all the referenced places

File: spanner/ddl/ast.go [hunk 4]
```diff
@@ -447,11 +447,15 @@ func GetSortedTableIdsBySpName(s Schema) []string {
 		tableName := tableQueue[0]
 		table := s[tableNameIdMap[tableName]]
 		tableQueue = tableQueue[1:]
+		parentTableExists := false
```
[the remark this answers is not in the document]
Comment (shreyakhajanchi) [reply spanner/ddl/ast.go:467]: done

**Review by bharadwaj-aditya — APPROVED** [review]
LGTM

**Review by bharadwaj-aditya — APPROVED** [review]

**Pull request merged**